pageserver/src/layered_repository/page_versions.rs

```diff
@@ -98,6 +98,24 @@ impl PageVersions {
         }
     }
 
+    pub fn read_pv(&self, off: u64) -> Result<PageVersion> {
+        let mut reader = self.reader(off)?;
+        Ok(PageVersion::des_from(&mut reader)?)
+    }
+
+    pub fn read_blob(&self, off: u64, buf: &mut Vec<u8>) -> Result<usize> {
+        // read length
+        let mut lenbuf = [0u8; 4];
+        self.file.read_exact_at(&mut lenbuf, off)?;
+        let len = u32::from_ne_bytes(lenbuf) as usize;
+
+        if buf.len() < len {
+            buf.resize(len, 0);
+        }
+        self.file.read_exact_at(&mut buf[0..len], off + 4)?;
+        Ok(len)
```

Comment:
This API might be simpler if we always resize `buf` to `len` and avoid returning `len`.

Reply:
Extending a vec requires initializing the new bytes, which isn't totally free. If we shrink the buffer, then we often need to extend it again on the next call. Maybe it's a premature optimization, I didn't try measuring it, but this is a pretty hot codepath..

Reply:
Added comments on that though.